File: templates/python/swimmer.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""Swimmer plot 患者时间线 —— 卡片 11_医学流行病学常用图/Swimmer图_Swimmer.md
每名患者一条横条=随访时长，叠加事件标记（应答/进展/死亡）+ 进行中箭头。RWD/肿瘤常用。
合成患者数据，可独立运行。改 # >>> PARAM 区换数据。"""
import os, sys, warnings; warnings.filterwarnings("ignore")
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
import numpy as np, pandas as pd, matplotlib.pyplot as plt
from _common import set_theme, save_fig, pal
set_theme(); np.random.seed(7)

# >>> PARAM ------------------------------------------------------
NAME = "tpl_swimmer"
TITLE = "Swimmer plot: patient timelines"; XLAB = "Months since enrollment"
BAR = pal("cat_main")[1]; C = pal("cat_main")
FIGSIZE = (4.6, 3.8)
# 换真实数据： 每行一患者：duration / response / progression / death(bool) / ongoing(bool)
m = 16
df = pd.DataFrame({
    "pid": [f"P{i:02d}" for i in range(1, m + 1)],
    "duration": np.round(np.random.uniform(3, 30, m), 1),
    "response": np.round(np.random.uniform(1, 8, m), 1),
    "progression": [np.round(np.random.uniform(8, 20), 1) if np.random.rand() < .6 else np.nan for _ in range(m)],
    "death": np.random.rand(m) < .35,
    "ongoing": np.random.rand(m) < .4})
# <<< PARAM ------------------------------------------------------

df = df.sort_values("duration").reset_index(drop=True)
fig, ax = plt.subplots(figsize=FIGSIZE)
for i, r in df.iterrows():
    ax.barh(i, r.duration, height=.6, color=BAR, alpha=.55, zorder=1)
    ax.scatter(r.response, i, marker="^", s=28, color=C[2], zorder=3)          # 应答
    if not np.isnan(r.progression) and r.progression <= r.duration:
        ax.scatter(r.progression, i, marker="X", s=30, color=C[5], zorder=3)   # 进展
    if r.death:
        ax.scatter(r.duration, i, marker="s", s=26, color="#333333", zorder=3) # 死亡
    elif r.ongoing:
        ax.annotate("", xy=(r.duration + 1.2, i), xytext=(r.duration, i),
                    arrowprops=dict(arrowstyle="-|>", color="grey", lw=1))      # 进行中
ax.set_yticks(range(len(df))); ax.set_yticklabels(df.pid, fontsize=6)
ax.set_xlabel(XLAB); ax.set_title(TITLE); ax.grid(False, axis="y")
handles = [plt.Line2D([], [], marker=m, ls="", color=c, label=l) for m, c, l in
           [("^", C[2], "Response"), ("X", C[5], "Progression"), ("s", "#333333", "Death")]]
handles.append(plt.Line2D([], [], marker=">", ls="", color="grey", label="Ongoing"))
ax.legend(handles=handles, fontsize=7, loc="lower right")
save_fig(fig, NAME)

File: templates/r/swimmer.R
## Swimmer plot 患者时间线 —— 卡片 11_医学流行病学常用图/Swimmer图_Swimmer.md
## 每名患者一条横条=随访时长 + 事件标记（应答/进展/死亡）+ 进行中箭头。
## 合成患者数据，可独立运行： Rscript templates/r/swimmer.R
.args <- commandArgs(trailingOnly = FALSE)
.sd <- dirname(normalizePath(sub("^--file=", "", .args[grep("^--file=", .args)])))
source(file.path(.sd, "_common.R"))
set.seed(2026)

# >>> PARAM ------------------------------------------------------
NAME <- "tpl_swimmer"; TITLE <- "Swimmer plot: patient timelines"; XLAB <- "Months since enrollment"
BAR <- pal("cat_main")[2]; C <- pal("cat_main")
# 换真实数据： 每行一患者：pid/duration/response/progression/death(逻辑)/ongoing(逻辑)
m <- 16
df <- data.frame(
  pid = sprintf("P%02d", 1:m), duration = round(runif(m, 3, 30), 1),
  response = round(runif(m, 1, 8), 1),
  progression = ifelse(runif(m) < .6, round(runif(m, 8, 20), 1), NA),
  death = runif(m) < .35, ongoing = runif(m) < .4)
# <<< PARAM ------------------------------------------------------

df <- df[order(df$duration), ]; df$y <- seq_len(nrow(df))
df$pid <- factor(df$pid, levels = df$pid)
prog <- df[!is.na(df$progression) & df$progression <= df$duration, ]
dead <- df[df$death, ]; ong <- df[!df$death & df$ongoing, ]
p <- ggplot(df) +
  geom_segment(aes(x = 0, xend = duration, y = y, yend = y), color = BAR, alpha = .55, linewidth = 3) +
  geom_point(aes(response, y), shape = 17, size = 2, color = C[3]) +                    # 应答
  { if (nrow(prog)) geom_point(data = prog, aes(progression, y), shape = 4, size = 2, stroke = 1.1, color = C[6]) } +
  { if (nrow(dead)) geom_point(data = dead, aes(duration, y), shape = 15, size = 2, color = "#333333") } +
  { if (nrow(ong))  geom_segment(data = ong, aes(x = duration, xend = duration + 1.5, y = y, yend = y),
                                 arrow = arrow(length = unit(1.6, "mm")), color = "grey40") } +
  scale_y_continuous(breaks = df$y, labels = df$pid) +
  labs(x = XLAB, y = NULL, title = TITLE) +
  theme(panel.grid.major.y = element_blank())
save_gg(p, NAME, w = 120, h = 100)
